lib/httpapimachinery/pkg/handler/stream.go

```diff
@@ -0,0 +1,107 @@
+package handler
+
+import (
+	"encoding/json"
+	"fmt"
+	_ctx "github.com/projectcalico/calico/lib/httpapimachinery/pkg/context"
+	"github.com/projectcalico/calico/lib/httpapimachinery/pkg/header"
+	"github.com/sirupsen/logrus"
+	"net/http"
+
+	"github.com/projectcalico/calico/lib/httpapimachinery/pkg/encoding"
+)
+
+// EventStream represents a http stream. Users can write objects to the response one at a time and the objects
+// will be flushed based on the configuration of the stream in the format configured in the stream.
+type EventStream[E any] interface {
+	Data(E) error
+	Error(msg string) error
+}
+
+type Event struct {
+	Type string `json:"type,omitempty"`
+	Data any    `json:"data"`
+}
+
+// jsonStreamWriter implements a ResponseStream. The format written to the http stream is json.
+type jsonEventStreamWriter[E any] struct {
+	*flusherWriter
+}
+
+func (w *jsonEventStreamWriter[E]) Error(msg string) error {
+	_, err := fmt.Fprintf(w, "error: %s\n\n", msg)
+	if err != nil {
+		return fmt.Errorf("failed to send event: %w", err)
+	}
+
+	w.Flush()
+	return nil
+}
+
+// Data encodes the given object as json and flushes it down the http stream.
+func (w *jsonEventStreamWriter[E]) Data(e E) error {
```

Comment:
Maybe `WriteData`? or just `Write`? e.g., `stream.Write(data)` - Might just be me, but `Data` doesn't imply an action is being take - rather I'd expect it to return me some Data.